skyportal/models.py

```diff
@@ -183,19 +183,78 @@ def desi_dr8_url(self):
                 f"&dec={self.dec}&size=200&layer=dr8&pixscale=0.262&bands=grz")
 
 
-GroupSource = join_model('group_sources', Group, Source)
+class Filter(Base):
+    query_string = sa.Column(sa.String, nullable=False, unique=False)
+    groups = relationship("Group", secondary="group_filters", back_populates="filters")
+
+
+GroupFilter = join_model("group_filters", Group, Filter)
+
+Candidate = join_model("candidates", Filter, Obj)
+Candidate.created_at = sa.Column(sa.DateTime, nullable=False, default=datetime.now,
+                                 index=True)
+Candidate.modified = sa.Column(sa.DateTime, nullable=False,
+                               server_default=sa.func.now(),
+                               server_onupdate=sa.func.now())
+Candidate.passing_alert_id = sa.Column(sa.Integer)
+
+
+def candidate_is_owned_by(self, user_or_token):
+    return bool(set(self.filter.groups) & set(user_or_token.groups))
+
+
+Candidate.is_owned_by = candidate_is_owned_by
+
+
+Source = join_model("sources", Group, Obj)
 """User.sources defines the logic for whether a user has access to a source;
    if this gets more complicated it should become a function/`hybrid_property`
    rather than a `relationship`.
 """
-User.sources = relationship('Source', backref='users',
-                            secondary='join(Group, group_sources).join(group_users)',
+Source.saved_by_id = sa.Column(sa.ForeignKey("users.id"), nullable=True, unique=False)
+Source.saved_by = relationship("User", foreign_keys=[Source.saved_by_id],
+                               backref="saved_sources")
+Source.saved_at = sa.Column(sa.DateTime, nullable=False, default=datetime.now)
+Source.active = sa.Column(sa.Boolean, default=True)
+Source.requested = sa.Column(sa.Boolean)
```

Comment:
@dannygoldstein I just copied this from your diagram, but don't know what it means; should it be nullable? If not, what should it default to?

Reply:
These columns implement the transfer / share logic described in #413 , 

`requested` should `server_default` to `False`, `active` should `server_default` to `True`, and neither should be nullable

> * Having an entry in the group_sources table prevents an object from being auto-deleted from the sidereal objects table (auto-candidate deletion).
>
> * If an object is saved to a group and then removed from the group, the corresponding group_sources entry is not deleted, but instead marked inactive to prevent the object from being demoted to candidate status and deleted, cascading the deletion to follow-up data
>
> * It is possible to transfer or share an object from one group to another, in the case of transferring, the group_sources entry is marked inactive and then a new group_sources entry is created for the proposed group, and marked requested, then someone from the other group has to accept the request, at which point the status is set to active and the source becomes visible to the group. in the case of sharing, everything happens as described in the previous sentence but the original group_sources entry is not marked inactive

Reply:
`saved_at` should be nullable, to allow manual posting of sources from `ATels`, `GCNs`, etc., and of other sources that do not start out as candidates from a datastream

Reply:
No need to have `modified` anymore if https://github.com/cesium-ml/baselayer/pull/60 is merged

Reply:
any reason to use DB-side `server_default` instead of just `default`?

Reply:
Just a convenience, saves time in case the user has to manually insert a row into the database from e.g., `psql`